babel.config.js

```diff
@@ -25,99 +25,394 @@
+    localPlugin('transform-fix-leading-comments'),
+    babelPlugin('transform-react-constant-elements'),
+    [babelPlugin('transform-classes'), {loose: false}],
+    [babelPlugin('transform-react-jsx'), defaultJsxOpts],
   ];
+  const unminifiedPresets = [defaultPresetEnvPlugin];
+  return {
+    compact: false,
+    ignore: ignoredGlobalModules,
+    plugins: unminifiedPlugins,
+    presets: unminifiedPresets,
+    sourceType: 'module',
+  };
+}
 
-  if (esm) {
-    return [leadingComments, reactConstantElements, transformJSX];
+/**
+ * Gets the config for pre-closure babel transforms run during `gulp dist`.
+ *
+ * @return {!Object}
+ */
+function getPreClosureConfig() {
+  const preClosurePlugins = [
+    localPlugin('transform-fix-leading-comments'),
+    babelPlugin('transform-react-constant-elements'),
+    [babelPlugin('transform-react-jsx'), defaultJsxOpts],
+    localPlugin('transform-inline-configure-component'),
+    // TODO(alanorozco): Remove `replaceCallArguments` once serving infra is up.
+    [localPlugin('transform-log-methods'), {replaceCallArguments: false}],
+    localPlugin('transform-parenthesize-expression'),
+    localPlugin('is_minified-constant-transformer'),
+    localPlugin('transform-amp-extension-call'),
+    localPlugin('transform-html-template'),
+    localPlugin('transform-version-call'),
+    localPlugin('transform-simple-array-destructure'),
+    replacePlugin,
+  ];
+  if (argv.single_pass) {
+    preClosurePlugins.push(localPlugin('transform-amp-asserts'));
+  }
+  if (argv.esm) {
+    preClosurePlugins.push(
+      defaultFilterImportsPlugin,
+      localPlugin('transform-function-declarations')
+    );
+  }
+  const isCheckTypes = argv._.includes('check-types');
+  if (isCheckTypes) {
+    preClosurePlugins.push(localPlugin('transform-simple-object-destructure'));
+  } else {
+    preClosurePlugins.push(localPlugin('transform-json-configuration'));
+  }
+  if (!argv.fortesting && !isCheckTypes) {
+    preClosurePlugins.push(
+      [localPlugin('amp-mode-transformer'), {isEsmBuild: argv.esm}],
+      localPlugin('is_dev-constant-transformer')
+    );
   }
+  const babelPresetEnvOptions = argv.esm
+    ? {bugfixes: true, modules: false, targets: {esmodules: true}}
+    : {
+        bugfixes: true,
+        loose: true,
+        modules: false,
+        targets: {'browsers': ['Last 2 versions']},
+      };
+  const preClosurePresets = [['@babel/preset-env', babelPresetEnvOptions]];
+  const preClosureConfig = {
+    compact: false,
+    plugins: preClosurePlugins,
+    presets: preClosurePresets,
+    retainLines: true,
+  };
+  if (argv.esm) {
+    preClosureConfig.sourceType = 'module';
+  }
+  return preClosureConfig;
+}
 
-  return [
-    leadingComments,
-    reactConstantElements,
-    transformClasses,
-    transformJSX,
-  ];
-};
+/**
+ * Gets the config for post-closure babel transforms run during `gulp dist`.
+ *
+ * @return {!Object}
+ */
+function getPostClosureConfig() {
```

Comment:
Since we're not intending to transform outside of the esm mode right now, should this instead return null or similar?